gnovm/cmd/gno/lint.go

```diff
@@ -140,6 +164,67 @@ func execLint(cfg *lintCfg, args []string, io commands.IO) error {
 	return nil
 }
 
+func lintTypeCheck(io commands.IO, memPkg *std.MemPackage, testStore gno.Store) (errorsFound bool, err error) {
+	tcErr := gno.TypeCheckMemPackageTest(memPkg, testStore)
+	if tcErr == nil {
+		return false, nil
+	}
+
+	errs := multierr.Errors(tcErr)
+	for _, err := range errs {
+		switch err := err.(type) {
+		case types.Error:
+			io.ErrPrintln(lintIssue{
+				Code:       lintTypeCheckError,
+				Msg:        err.Msg,
+				Confidence: 1,
+				Location:   err.Fset.Position(err.Pos).String(),
+			})
+		case scanner.ErrorList:
+			for _, scErr := range err {
+				io.ErrPrintln(lintIssue{
+					Code:       lintParserError,
+					Msg:        scErr.Msg,
+					Confidence: 1,
+					Location:   scErr.Pos.String(),
+				})
+			}
+		case scanner.Error:
+			io.ErrPrintln(lintIssue{
+				Code:       lintParserError,
+				Msg:        err.Msg,
+				Confidence: 1,
+				Location:   err.Pos.String(),
+			})
+		default:
+			//nolint
+			return false, fmt.Errorf("unexpected error type: %T", err)
+		}
+	}
+	return true, nil
+}
+
+func lintTestFiles(memPkg *std.MemPackage) *gno.FileSet {
+	testfiles := &gno.FileSet{}
+	for _, mfile := range memPkg.Files {
+		if !strings.HasSuffix(mfile.Name, ".gno") {
+			continue // Skip non-GNO files
+		}
+
+		n, _ := gno.ParseFile(mfile.Name, mfile.Body)
+		if n == nil {
+			continue // Skip empty files
+		}
+
+		// XXX: package ending with `_test` is not supported yet
+		if strings.HasSuffix(mfile.Name, "_test.gno") && !strings.HasSuffix(string(n.PkgName), "_test") {
+			// Keep only test files
+			testfiles.AddFiles(n)
+		}
```

Comment:
This code block should be moved *before* `gno.ParseFile(...)` line 214.
I saw somewhere else in this PR that there is also some code to match against `*_filetest.gno`. Should it be added here, or better, get rid of special treatment for `*_filetest.gno`. Why should it be different from `*_test.gno`?

Reply:
We need to find n.PkgName which is only available after parse. I agree that the code is not ideal, but this is mostly the result of a code move (see old code)

Filetest work very differently from tests, so should be excluded and should not even be linted for now... (possibly removed: #1346)